client_configuration_finder.go

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2023 Snowflake Computing Inc. All rights reserved.
+
+package gosnowflake
+
+import (
+	"errors"
+	"os"
+	"path"
+)
+
+const (
+	defaultConfigName = "sf_client_config.json"
+	clientConfEnvName = "SF_CLIENT_CONFIG_FILE"
+)
+
+func getClientConfig(filePathFromConnectionString string) (*ClientConfig, error) {
+	configPredefinedFilePaths, err := clientConfigPredefinedDirs()
+	var filePath string
+	filePath, err = findClientConfig(filePathFromConnectionString, configPredefinedFilePaths)
+	if err != nil {
+		return nil, err
+	}
+	if filePath == "" { // we did not find a config file
+		return nil, nil
+	}
+	return parseClientConfiguration(filePath)
+}
+
+func findClientConfig(filePathFromConnectionString string, configPredefinedDirs []string) (string, error) {
+	if filePathFromConnectionString != "" {
+		return filePathFromConnectionString, nil
+	}
+	envConfigFilePath := os.Getenv(clientConfEnvName)
+	if envConfigFilePath != "" {
+		return envConfigFilePath, nil
+	}
+	return searchForConfigFile(configPredefinedDirs)
+}
+
+func searchForConfigFile(directories []string) (string, error) {
+	for _, dir := range directories {
+		filePath := path.Join(dir, defaultConfigName)
+		exists, err := existsFile(filePath)
+		if err != nil {
```

Comment:
Is it ok? Should we return as early error or should we proceed to the next available directory and return error(s) at the end? How did you implemented it in other drivers?

Reply:
if IO operation fails on unexpected error - different than file not exists, then operation fails.
That's how it was done in other drivers. Do you think that we should just treat it as if the file did not exist?

Reply:
That is a very good question. If first directory is not usable (for instance not readable because it is run in a specific docker env?) it blocks other directories - so blocks the feature at all.